### ISSUE TITLE
WARNING cannot check MD5 signature since it was unset in the STREAMINFO

### ISSUE THREAD

**Issue by klslz** [xiph/flac#155]
To run integrity checks I use respective functions:

```
flac --test test.flac
flac --analyze test.flac
```
these are giving me:

```
flac 1.3.3-b84ff55
Copyright (C) 2000-2009  Josh Coalson, 2011-2019  Xiph.Org Foundation
flac comes with ABSOLUTELY NO WARRANTY.  This is free software, and you are
welcome to redistribute it under certain conditions.  Type `flac' for details.

test.flac: WARNING, cannot check MD5 signature since it was unset in the STREAMINFO
ok
```
               
What does it mean? It's "ok" (probably the decoding part) but It can not check the MD5 signature?
Does  this work as integrity check at all!?!?

And I do not have it on all flacs though!
I have it now on newly converted wav and also reencoded flacs.

The flac git version acts the same way as the official 1.3.3. version I have on my Ubuntu. 

Any hint is a appreciated.

**Comment by rillian** [comment]
This warning [means](https://github.com/xiph/flac/blob/edfbe3cf1947/src/flac/decode.c#L491-L495) the encoder didn't record a checksum in the [stream header](https://xiph.org/flac/format.html#metadata_block_streaminfo).

The `ok` means the decoder was able to decode all the data, but because the original encoder didn't record an MD5 sum, it's couldn't compare the decoded data to check that it was bit-identical to the original. This might happen, for example, if the flac was recorded live, so the creating application couldn't know what the complete audio stream would be.

In other words, the files aren't corrupt, but have weaker integrity guarantees. Re-encoding is a good way to address the issue.

**Comment by klslz** [comment]
I tried both: 

pcm>>flac and 
flac>>flac

The message persists.

**Comment by rillian** [comment]
That's more surprising. Can you share an example of the file that's failing?

I created a test file by re-encoding with `--no-md5-sum`:
```
$ flac song.flac --no-md5-sum -o test.flac
[...]
song.flac: WARNING, MD5 computation disabled, resulting file will not have MD5 sum
song.flac: wrote 25481029 bytes, ratio=0.999
```

Re-encoding the test file restores the checksum:
```
$ flac test.flac -o reencode.flac
[...]
test.flac: wrote 25481029 bytes, ratio=1.000

$ flac --test *.flac
[...]
reencode.flac: ok                    
song.flac: ok                    
test.flac: WARNING, cannot check MD5 signature since it was unset in the STREAMINFO
ok   
```

I see the same results with flac 1.3.2 (Fedora 30) and a build of current git master.

**Comment by klslz** [comment]
I tested it now on my RPI. That worked fine!
This only happens on less then 1% of my flacs on my recently upgraded Ubuntu machine.
I copied these files on a tmpfs. Didin't work. And the same files I copied over to the RPI. And that worked. Weird. I also reinstalled libflac and flac.

I even converted the flacs with sox to RAW and then back to flac and it still happened.
Hmmh. I guess it requires more digging.

**Comment by klslz** [comment]
Update:

I guess my system was broken. I reinstalled my WS (I now switched to Fedora) and retested the affected files.

All files seem to be OK now. Which was pretty much expected after the first test on my RPI.

Thx for the support. And sorry for the confusion.

Case closed.

**Comment by rillian** [comment]
Glad you got it working in the end!

**Comment by klslz** [comment]
It was still an issue on my other platforms. 


As outlined at the end of #192 
There's been a patch developed.

It turned out to be gcc related.
https://sourceforge.net/p/flac/bugs/478/

I applied the patch and now all is fine.

**Comment by magicgoose** [comment]
I have the same issue.
All files which are affected have 1 thing in common:
**the audio MD5 (if presented as hex string) starts with `00`.**

**Comment by a-vrma** [comment]
Pretty sure this is the same bug as #192.

> Anybody issued a pull request?

Yes, pull request #227

**Comment by ktmf01** [comment]
I presume this was fixed by #227, which was included in release 1.3.4, therefore I'm closing this issue.

**Comment by lamyergeier** [comment]
@ktmf01 Still facing it in version 1.3.4! On linuxmint 20.3 machine

**Comment by ktmf01** [comment]
Is it the exact same issue? Can you post the output of `metaflac --list` on that file?

**Comment by lamyergeier** [comment]
```
METADATA block #0
  type: 0 (STREAMINFO)
  is last: false
  length: 34
  minimum blocksize: 4608 samples
  maximum blocksize: 4608 samples
  minimum framesize: 16 bytes
  maximum framesize: 13353 bytes
  sample_rate: 192000 Hz
  channels: 2
  bits-per-sample: 24
  total samples: 231761920
  MD5 signature: 00000000000000000000000000000000
METADATA block #1
  type: 3 (SEEKTABLE)
  is last: false
  length: 2178
  seek points: 121
    point 0: sample_number=0, stream_offset=0, frame_samples=4608
    point 1: sample_number=1921536, stream_offset=2494190, frame_samples=4608
    point 2: sample_number=3843072, stream_offset=5366318, frame_samples=4608
    point 3: sample_number=5760000, stream_offset=8279420, frame_samples=4608
    point 4: sample_number=7681536, stream_offset=11060926, frame_samples=4608
    point 5: sample_number=9603072, stream_offset=13906991, frame_samples=4608
    point 6: sample_number=11520000, stream_offset=16770607, frame_samples=4608
    point 7: sample_number=13441536, stream_offset=19679007, frame_samples=4608
    point 8: sample_number=15363072, stream_offset=22573054, frame_samples=4608
    point 9: sample_number=17280000, stream_offset=25447990, frame_samples=4608
    point 10: sample_number=19201536, stream_offset=28316912, frame_samples=4608
    point 11: sample_number=21123072, stream_offset=31096923, frame_samples=4608
    point 12: sample_number=23040000, stream_offset=33905052, frame_samples=4608
    point 13: sample_number=24961536, stream_offset=36727088, frame_samples=4608
    point 14: sample_number=26883072, stream_offset=39563403, frame_samples=4608
    point 15: sample_number=28800000, stream_offset=42467104, frame_samples=4608
    point 16: sample_number=30721536, stream_offset=45447317, frame_samples=4608
    point 17: sample_number=32643072, stream_offset=48307805, frame_samples=4608
    point 18: sample_number=34560000, stream_offset=51159272, frame_samples=4608
    point 19: sample_number=36481536, stream_offset=54071559, frame_samples=4608
    point 20: sample_number=38403072, stream_offset=56990012, frame_samples=4608
    point 21: sample_number=40320000, stream_offset=59907585, frame_samples=4608
    point 22: sample_number=42241536, stream_offset=62804610, frame_samples=4608
    point 23: sample_number=44163072, stream_offset=65768293, frame_samples=4608
    point 24: sample_number=46080000, stream_offset=68834590, frame_samples=4608
    point 25: sample_number=48001536, stream_offset=71830207, frame_samples=4608
    point 26: sample_number=49923072, stream_offset=74668427, frame_samples=4608
    point 27: sample_number=51840000, stream_offset=77439355, frame_samples=4608
    point 28: sample_number=53761536, stream_offset=80266360, frame_samples=4608
    point 29: sample_number=55683072, stream_offset=83065419, frame_samples=4608
    point 30: sample_number=57600000, stream_offset=86015062, frame_samples=4608
    point 31: sample_number=59521536, stream_offset=88986525, frame_samples=4608
    point 32: sample_number=61443072, stream_offset=91990205, frame_samples=4608
    point 33: sample_number=63360000, stream_offset=94978641, frame_samples=4608
    point 34: sample_number=65281536, stream_offset=97961136, frame_samples=4608
    point 35: sample_number=67203072, stream_offset=100971109, frame_samples=4608
    point 36: sample_number=69120000, stream_offset=103948200, frame_samples=4608
    point 37: sample_number=71041536, stream_offset=106862162, frame_samples=4608
    point 38: sample_number=72963072, stream_offset=109794322, frame_samples=4608
    point 39: sample_number=74880000, stream_offset=112779360, frame_samples=4608
    point 40: sample_number=76801536, stream_offset=115708788, frame_samples=4608
    point 41: sample_number=78723072, stream_offset=118691851, frame_samples=4608
    point 42: sample_number=80640000, stream_offset=121628647, frame_samples=4608
    point 43: sample_number=82561536, stream_offset=124596878, frame_samples=4608
    point 44: sample_number=84483072, stream_offset=127609965, frame_samples=4608
    point 45: sample_number=86400000, stream_offset=130570526, frame_samples=4608
    point 46: sample_number=88321536, stream_offset=133452940, frame_samples=4608
    point 47: sample_number=90243072, stream_offset=136437451, frame_samples=4608
    point 48: sample_number=92160000, stream_offset=139417094, frame_samples=4608
    point 49: sample_number=94081536, stream_offset=142497581, frame_samples=4608
    point 50: sample_number=96003072, stream_offset=145475391, frame_samples=4608
    point 51: sample_number=97920000, stream_offset=148517136, frame_samples=4608
    point 52: sample_number=99841536, stream_offset=151517926, frame_samples=4608
    point 53: sample_number=101763072, stream_offset=154506654, frame_samples=4608
    point 54: sample_number=103680000, stream_offset=157505210, frame_samples=4608
    point 55: sample_number=105601536, stream_offset=160473182, frame_samples=4608
    point 56: sample_number=107523072, stream_offset=163381281, frame_samples=4608
    point 57: sample_number=109440000, stream_offset=166228458, frame_samples=4608
    point 58: sample_number=111361536, stream_offset=169115616, frame_samples=4608
    point 59: sample_number=113283072, stream_offset=171993400, frame_samples=4608
    point 60: sample_number=115200000, stream_offset=174808430, frame_samples=4608
    point 61: sample_number=117121536, stream_offset=177605277, frame_samples=4608
    point 62: sample_number=119043072, stream_offset=180396593, frame_samples=4608
    point 63: sample_number=120960000, stream_offset=183343648, frame_samples=4608
    point 64: sample_number=122881536, stream_offset=186362272, frame_samples=4608
    point 65: sample_number=124803072, stream_offset=189374675, frame_samples=4608
    point 66: sample_number=126720000, stream_offset=192403930, frame_samples=4608
    point 67: sample_number=128641536, stream_offset=195425168, frame_samples=4608
    point 68: sample_number=130563072, stream_offset=198532015, frame_samples=4608
    point 69: sample_number=132480000, stream_offset=201567297, frame_samples=4608
    point 70: sample_number=134401536, stream_offset=204505075, frame_samples=4608
    point 71: sample_number=136323072, stream_offset=207521316, frame_samples=4608
    point 72: sample_number=138240000, stream_offset=210573286, frame_samples=4608
    point 73: sample_number=140161536, stream_offset=213554602, frame_samples=4608
    point 74: sample_number=142083072, stream_offset=216494148, frame_samples=4608
    point 75: sample_number=144000000, stream_offset=219378288, frame_samples=4608
    point 76: sample_number=145921536, stream_offset=222306466, frame_samples=4608
    point 77: sample_number=147843072, stream_offset=225318678, frame_samples=4608
    point 78: sample_number=149760000, stream_offset=228397611, frame_samples=4608
    point 79: sample_number=151681536, stream_offset=231419036, frame_samples=4608
    point 80: sample_number=153603072, stream_offset=234445570, frame_samples=4608
    point 81: sample_number=155520000, stream_offset=237456622, frame_samples=4608
    point 82: sample_number=157441536, stream_offset=240540414, frame_samples=4608
    point 83: sample_number=159363072, stream_offset=243624968, frame_samples=4608
    point 84: sample_number=161280000, stream_offset=246621120, frame_samples=4608
    point 85: sample_number=163201536, stream_offset=249672831, frame_samples=4608
    point 86: sample_number=165123072, stream_offset=252766194, frame_samples=4608
    point 87: sample_number=167040000, stream_offset=255866758, frame_samples=4608
    point 88: sample_number=168961536, stream_offset=258969253, frame_samples=4608
    point 89: sample_number=170883072, stream_offset=262051148, frame_samples=4608
    point 90: sample_number=172800000, stream_offset=265099252, frame_samples=4608
    point 91: sample_number=174721536, stream_offset=268070966, frame_samples=4608
    point 92: sample_number=176643072, stream_offset=270936314, frame_samples=4608
    point 93: sample_number=178560000, stream_offset=273911961, frame_samples=4608
    point 94: sample_number=180481536, stream_offset=276934977, frame_samples=4608
    point 95: sample_number=182403072, stream_offset=280009808, frame_samples=4608
    point 96: sample_number=184320000, stream_offset=283058607, frame_samples=4608
    point 97: sample_number=186241536, stream_offset=286134448, frame_samples=4608
    point 98: sample_number=188163072, stream_offset=289206469, frame_samples=4608
    point 99: sample_number=190080000, stream_offset=292238838, frame_samples=4608
    point 100: sample_number=192001536, stream_offset=295256487, frame_samples=4608
    point 101: sample_number=193923072, stream_offset=298279984, frame_samples=4608
    point 102: sample_number=195840000, stream_offset=301262010, frame_samples=4608
    point 103: sample_number=197761536, stream_offset=304244637, frame_samples=4608
    point 104: sample_number=199683072, stream_offset=307174001, frame_samples=4608
    point 105: sample_number=201600000, stream_offset=310015922, frame_samples=4608
    point 106: sample_number=203521536, stream_offset=312863747, frame_samples=4608
    point 107: sample_number=205443072, stream_offset=315740646, frame_samples=4608
    point 108: sample_number=207360000, stream_offset=318607332, frame_samples=4608
    point 109: sample_number=209281536, stream_offset=321459816, frame_samples=4608
    point 110: sample_number=211203072, stream_offset=324294225, frame_samples=4608
    point 111: sample_number=213120000, stream_offset=327111684, frame_samples=4608
    point 112: sample_number=215041536, stream_offset=329983354, frame_samples=4608
    point 113: sample_number=216963072, stream_offset=332900496, frame_samples=4608
    point 114: sample_number=218880000, stream_offset=335782696, frame_samples=4608
    point 115: sample_number=220801536, stream_offset=338681163, frame_samples=4608
    point 116: sample_number=222723072, stream_offset=341516812, frame_samples=4608
    point 117: sample_number=224640000, stream_offset=344361686, frame_samples=4608
    point 118: sample_number=226561536, stream_offset=347174142, frame_samples=4608
    point 119: sample_number=228483072, stream_offset=349961059, frame_samples=4608
    point 120: sample_number=230400000, stream_offset=352739209, frame_samples=4608
METADATA block #2
  type: 4 (VORBIS_COMMENT)
  is last: false
  length: 632
  vendor string: Mutagen 1.42.1
  comments: 13
    comment[0]: ARTIST=Wilhelm Furtwängler
    comment[1]: COMPOSER=Pyotr Illitch Tchaïkovski
    comment[2]: COPYRIGHT=A Warner Classics release, © 2021 Parlophone Records Limited A Warner Classics release, ℗ 1939 Parlophone Records Limited. Remastered 2020 Parlophone Records Limited
    comment[3]: TITLE=Symphony No. 6 in B Minor, Op. 74 "Pathétique": I. Adagio - Allegro non troppo
    comment[4]: ISRC=GBAYC2001983
    comment[5]: ALBUM=Tchaikovsky: Serenade for Strings, Symphonies Nos. 4 & 6 "Pathétique"
    comment[6]: TRACKTOTAL=11
    comment[7]: ALBUMARTIST=Wilhelm Furtwängler
    comment[8]: COMMENT=https://www.isrbx.net
    comment[9]: GENRE=Classique
    comment[10]: DATE=2021
    comment[11]: ORGANIZATION=Warner Classics
    comment[12]: TRACKNUMBER=01
METADATA block #3
  type: 1 (PADDING)
  is last: true
  length: 1322170
```

**Comment by ktmf01** [comment]
In your case `WARNING, cannot check MD5 signature since it was unset in the STREAMINFO` is actually correct: the MD5 is unset in this file, this is not a bug.

If you want to fix this, re-encode your file, which you can do with sox as you pointed out. You can also use `flac -f yourfile.flac` to re-encode

**Comment by lamyergeier** [comment]
@ktmf01 Before my first comment, I did the following:

```bash
flac --verify --decode-through-errors --preserve-modtime -o a1.flac a.flac
```
but 

the output shows the same error. The reencoding with flac doesn't help while sox fixes it!

**Comment by ktmf01** [comment]
Ok, that is a different issue then, the flac encoder not writing an MD5sum when writing to a file.

Are you sure you used the command you mentioned in your previous post though? The vendor string says Mutagen, not libFLAC, and your file has blocks of 4608 samples, while libFLAC uses 4096 by default. That doesn't seem to add up.

**Comment by lamyergeier** [comment]
I am sorry, I am not sure as I am not able to replicate the above results from metaflac command.  Please ignore my issue. It seems to be working fine now.
